csrc/val_graph.cpp

```diff
@@ -0,0 +1,476 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <id_model/to_string.h>
+#include <ir/utils.h>
+#include <val_graph.h>
+
+namespace nvfuser {
+
+namespace {
+using UnorderedSetOfExprGroup = std::unordered_set<ExprGroup>;
+using DequeOfExprGroup = std::deque<ExprGroup>;
+} // namespace
+
+ValGraph::ValGraph(const ValGraph& other)
+    : disjoint_vals_(other.disjoint_vals_),
+      disjoint_exprs_(other.disjoint_exprs_),
+      unique_definitions_(),
+      unique_uses_() {
+  for (const auto& [orig_val_group, orig_expr_groups] :
+       other.unique_definitions_) {
+    auto new_val_group = toGroup(orig_val_group->front());
+
+    ExprGroups new_expr_groups;
+    for (const ExprGroup& orig_expr_group : orig_expr_groups) {
+      new_expr_groups.pushBack(toGroup(orig_expr_group->front()));
+    }
+
+    unique_definitions_[new_val_group] = new_expr_groups;
```

Comment:
nit:
```suggestion
    unique_definitions_[new_val_group] = std::move(new_expr_groups);
```

Reply:
Changed to `unique_definitions_.emplace(new_val_group, std::move(new_expr_groups));` as just using `std::move` won't actually move (per clang-tidy)